### ISSUE TITLE
Memory leak in http server

### ISSUE THREAD

**Issue by kamalmarhubi** [alexcrichton/futures-rs#3]
Memory grows linearly with number of requests, observing about 1.5 K per request in growth.

The futures machinery makes it really hard to look at a heap profiler's output. Here's output from running
`valgrind --tool=massif target/release/http` (and compiling with `alloc_system`): https://gist.github.com/kamalmarhubi/377504123faef12365ca01e28f1e82ed

The 1.5 K per request is with either jemalloc or system allocator, and is quite linear.

**Comment by kamalmarhubi** [comment]
I don't think this is in the http server. A simple qotd server is growing about 500 bytes per request. Source: https://github.com/kamalmarhubi/futures-rs/blob/qotd/mio/examples/qotd.rs

**Comment by kamalmarhubi** [comment]
After @jvns [showed me ltrace last night](http://jvns.ca/blog/2016/06/15/using-ltrace-to-debug-a-memory-leak/), we tracked down some leaking allocations and found that in fact the very first `accept(listener)` allocation is never freed.

In both cases, the heap growth is from holding on to the initial `accept(listener)` future, which is `await`'ed. This future never completes: instead a chain of futures is created via `join`, which grows with each request received. Using `join` ensures we listen for the next request, but at the cost of making the *current* request's future not complete until the next request has. And so on.

I think the `pub fn accept(&self) -> Box<IoFuture<(TcpStream, SocketAddr)>>` API is what is "wrong" here: wrapping this up in a future is conceptually nice but has issues. I'm experimenting with a replacement listening API like this:

```
impl Loop {
    pub fn start_server<F>(&mut self, addr: &SocketAddr, cb: F)
                           -> io::Result<()>
        where F: Fn(TcpStream, SocketAddr) -> Box<IoFuture<()>> + 'static
}
```

This works for the qotd server, keeping memory usage steady over at least sequential requests. (This wasn't true previously!) For the HTTP server, sequential requests have memory usage steady, but concurrent requests just don't work right now. I think I messed something up in socket registration.
